test/Npgsql.Tests/PrepareTests.cs

```diff
@@ -70,6 +70,35 @@ public void Bug3443()
         Assert.That(cmd2.IsPrepared, Is.True);
     }
 
+    [Test, IssueLink("https://github.com/npgsql/npgsql/issues/4209")]
+    public async Task Async_cancel_NullReferenceException()
+    {
+        for (var i = 0; i < 10; i++)
+        {
+            using var conn = OpenConnectionAndUnprepare();
+            using var cmd = new NpgsqlCommand("SELECT 1", conn);
+            using var cts = new CancellationTokenSource();
+            using var mre = new ManualResetEventSlim();
+            var cancelTask = Task.Run(() =>
+            {
+                mre.Wait();
+                cts.Cancel();
+            });
+            try
+            {
+                mre.Set();
+                await cmd.PrepareAsync(cts.Token);
+            }
+            catch (OperationCanceledException)
+            {
+                // expected exception
```

Comment:
Can use Assert.ThrowsAsync

Reply:
Not really. The cancellation can happen before we check the token (in which case, `OCE` is thrown), or after (no exception).

Reply:
Right, makes sense